File: try/match.js
const { MongoClient } = require('mongodb');

// Connect to MongoDB
const uri = 'mongodb://localhost:27017/';
const client = new MongoClient(uri, { useNewUrlParser: true, useUnifiedTopology: true });

async function getData() {
    try {
        await client.connect();
        const database = client.db('new_web');
        const studentsCollection = database.collection('students');
        const positionsCollection = database.collection('positions');
        const promisesCollection = database.collection('promises');

        // Retrieve data from MongoDB collections
        const idGpa = await studentsCollection.find({}, { projection: { _id: 1, gpa: 1 } }).toArray();
        const positionsData = await positionsCollection.find({}, { projection: { _id: 1, capacity: 1, gpa_required: 1 } }).toArray();
        const promisesData = await promisesCollection.find({}, { projection: { student_id: 1, promise1: 1, promise2: 1, promise3: 1 } }).toArray();

        const students = idGpa.map(item1 => {
            const promises = promisesData.find(item2 => item1._id === item2.student_id);
            return [item1._id, item1.gpa, promises.promise1, promises.promise2, promises.promise3];
        });

        const ids = students.map(x => x[0]);
        const mvt = positionsData.map(x => x._id);

        console.log(ids);
        console.log(mvt);

        const promision = {};
        for (let i = 0; i < 3; i++) {
            const pm = {};
            positionsData.forEach(position => {
                students.forEach(student => {
                    if (student[2 + i] === position._id) {
                        if (!pm[position._id]) {
                            pm[position._id] = [];
                        }
                        pm[position._id].push([student[0], student[1]]);
                    }
                });
            });
            const lstPm = Object.entries(pm).map(([k, v]) => [k, v, ...positionsData.find(position => position._id === k).slice(1)]);
            const dctPm = Object.fromEntries(lstPm.map(x => [x[0], x.slice(1)]));
            promision[`NV${i + 1}`] = dctPm;
        }
        console.log(`Vị trí theo nguyện vọng: ${JSON.stringify(promision)}`);

        // Function to filter students based on GPA threshold
        function cut(lst, k) {
            return lst.filter(x => x[1] >= k).map(x => [x[0], x[1]]);
        }

        // Function to delete matched students from the list of IDs
        function deleteIds(ids, lst) {
            return ids.filter(id => lst === null || !lst.includes(id));
        }

        // Function for matching students to positions
        function matching(promision, ids, mvtList) {
            const result = {};
            for (const nv in promision) {
                const dct = promision[nv];
                for (const mvt of mvtList) {
                    const listStudent = dct[mvt];
                    listStudent.sort((a, b) => b[1] - a[1]);
                    listStudent = cut(listStudent, listStudent[2]);
                    if (listStudent.length > listStudent[1]) {
                        listStudent = listStudent.slice(0, listStudent[1]);
                    }
                    const msvs = listStudent.map(x => x[0]);
                    for (const msv of msvs) {
                        if (ids.includes(msv)) {
                            if (!result[mvt]) {
                                result[mvt] = [];
                            }
                            result[mvt].push(msv);
                            ids = deleteIds(ids, msv);
                        }
                    }
                }
            }
            return result;
        }

        const matchStd = matching(promision, ids, mvt);

        console.log('Kết quả matching:');
        console.log(matchStd);

        // Write the result to a file or perform further actions as needed
    } finally {
        await client.close();
    }
}

getData();

File: try/match.py
# from fastapi import FastAPI
# import mysql.connector

# host = 'localhost'
# user = 'root'
# password = ''
# database = 'thuctap'

# conn = mysql.connector.connect(host=host,
#                                user=user,
#                                password = password,
#                                database= database)
# list1 = []
# list2 = []
# positions = []
# cursor = conn.cursor()
# cursor.execute("SELECT masv,gpa from sinhvien")
# result = cursor.fetchall()
# for row in result:
#     list1.append((row[0], row[1]))

# cursor.execute("SELECT * from nguyenvong")
# result = cursor.fetchall()
# for row in result:
#     list2.append((row[0], row[1],row[2],row[3]))

# cursor.execute("SELECT mavt,diemsan, soluongtuyen from vitri")
# result = cursor.fetchall()
# for row in result:
#     positions.append((row[0], row[1],row[2]))
# # list1 = [(100, 3.0), (101, 3.2), (102, 3.3), (104, 2.8), (105, 2.9)]
# # list2 = [(100, 1, 2, 3), (101, 3, 2, 1), (102, 2, 1, 3), (103, 1, 3, 2), (104, 1, 2, 3), (105, 3, 2, 1)]
# # positions = [(1, 3.0, 1), (2, 3.0, 2), (3, 2.8, 2)]
# # print(f"Sinh viên : {list1}")
# # print()
# # print(f"Nguyện vọng: {list2}")
# # print()
# # print(f"Vị trí: {positions}")
# # print()
# students = [(item1[0], item1[1], *item2[1:]) for item1 in list1 for item2 in list2 if item1[0] == item2[0]]
# id = [x[0] for x in students]


id_gpa = []
with open("idAndGpa.txt") as f:
    for line in f:
        tokens = line.strip().split(", ")
        id_gpa.append((tokens[0],float(tokens[1])))

positions = []
with open("positions.txt") as f:
    for line in f:
        tokens = line.strip().split(", ")
        positions.append((tokens[0], float(tokens[1]), int(tokens[2])))

promises = []
with open("promises.txt") as f:
    for line in f:
        tokens = line.strip().split(", ")
        promises.append((tokens[0], tokens[1], tokens[2], tokens[3]))

students = [(item1[0], item1[1], *item2[1:]) for item1 in id_gpa for item2 in promises if item1[0] == item2[0]]
id = [x[0] for x in students]

promision = {}
for i in range(3):
    pm = {position[0] : [] for position in positions}
    for position in positions:
        for student in students:
            if(student[2+i] == position[0]):
                pm[position[0]].append((student[0],student[1]))
    lst_pm = [[k,v, *position[1:]] for k,v in pm.items() for position in positions if position[0] == k]
    dct_pm = {x[0] : x[1:] for x in lst_pm}
    promision[f'NV{i+1}'] = dct_pm
print(f"Vị trí theo nguyện vọng: {promision}")

#loại bỏ đi các sinh viên có điểm thấp hơn điểm sàn
def cut(lst, k):
    cut_list = []
    for x in lst:
        if(x[1] >= k):
            cut_list.append((x[0],x[1]))
    return cut_list
#sau khi match được sinh viên thì xóa sinh viên đó khỏi list id
def delete(id,lst):
    new_id = []
    for i in range(len(id)):
        if lst is not None and id[i] not in lst:
            new_id.append(id[i])
    return new_id
def matching(dct,id):
    result = {mvt : [] for mvt in dct.keys()}
    for mvt in dct.keys():
        p = dct[mvt]
        p[0].sort(key= lambda x: x[1], reverse = True)
        p[0] = cut(p[0],p[1])
        if(len(p[0]) > p[2]):
            p[0] = (p[0])[:(p[2])]
        msvs = [x[0] for x in p[0]]
        for msv in msvs:
            if msv in id:
                result[mvt].append(msv)
                id = delete(id,result[mvt])
    return result
dct = promision['NV1']
match_std = matching(dct,id)
with open("output_match.txt", "w", encoding="utf-8") as output_file:
    output_file.write("Kết quả matching:\n")
    for key, value in match_std.items():
        output_file.write(f"{key}: {value}\n")
print("Kết quả matching:")
print(match_std)

    

